tcp.rflx

```diff
@@ -0,0 +1,124 @@
+package TCP is
+
+    -- RFC 793
+
+    type Source_Port is mod 2**16;
+    type Destination_Port is mod 2**16;
+    type Sequence_Number is mod 2**32;
+    type Acknowledgment_Number is mod 2**32;
+    type Data_Offset is range 5 .. 2**4 - 1 with Size => 4;
+
+    -- Reserved for future use. Must be zero.
+    type Reserved is range 0 .. 0 with Size => 3;
+
+    type Window is mod 2**16;
+    type Checksum is mod 2**16;
+    type Urgent_Pointer is mod 2**16;
+
+    -- Timestamp and Window_Scale options are described in RFC 7323
+    -- Selective Acknowledgment Options (SACK) and SACK_Permitted are described in RFC 2018
+    -- No_Operation, End_Of_Options_List are described in RFC 793
+    -- Maximum_Segment_Size is described in RFC 793 and RFC 879
+
+    type Kind is
+       (End_Of_Options_List    => 16#00#,
+        No_Operation           => 16#01#,
+        Maximum_Segment_Size   => 16#02#,
+        Window_Scale           => 16#03#,
+        SACK_Permitted         => 16#04#,
+        SACK                   => 16#05#,
+        Timestamp              => 16#08#)
+    with Size => 8;
+
+    type Length is range 2 .. 2**8 - 1 with Size => 8;
+    type TS_Length is range 10 .. 10 with Size => 8;
+    type TSval is mod 2**32;
+    type TSecr is mod 2**32;
+    type Shift_cnt is mod 2**8;
+    type MaxSegSize is mod 2**16;
+
+
+    type Option is
+       message
+          Kind : Kind
+             then null
+                if Kind = End_Of_Options_List or Kind = No_Operation
+             then TS_Length
+                if Kind = Timestamp
+             then Length
+                if Kind = Maximum_Segment_Size
+                   or Kind = Window_Scale
+                   or Kind = SACK_Permitted
+                   or Kind = SACK;
+          TS_Length : TS_Length;
+          TSval : TSval
+                if Kind = Timestamp;
+          TSecr : TSecr
+                if Kind = Timestamp
+             then null
+                if Kind = Timestamp;
+          Length : Length
+             then Shift_cnt
+                if Kind = Window_Scale
+             then MaxSegSize
+                if Kind = Maximum_Segment_Size
+             then SACK_Data
+                if Kind = SACK
+             then null
+                if Kind = SACK_Permitted;
+          Shift_cnt : Shift_cnt
+             then null
+                if Kind = Window_Scale;
+          MaxSegSize : MaxSegSize
+             then null
+                if Kind = Maximum_Segment_Size;
+          SACK_Data : Opaque
+              with Size => Length * 8 - Kind'Size - Length'Size
+             then null
+                if Kind = SACK;
```

Comment:
The relation between `Length` and the actual message size should be defined by adding `Message'Size = Length * 8` to all link conditions to the final field (`null`). This will allow the verifier to find all missing value constraints for `Lenght`. In the long term we want to use a global invariant for such properties (cf. https://github.com/Componolit/RecordFlux/issues/554). Please add a reference to the corresponding issue: `--  ISSUE: Componolit/RecordFlux#554`.

Reply:
This relation is added in the commit 17a607e6f9a09634d1cb20ee17e2f885e2024512.

Reply:
Done in commit b040f2e19b89c3d5e955c08db220ad93426d4e19.